File: spriteforest-x/extension/SFTouchSprite.cpp
//
//  SFTouchSprite.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-27.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFTouchSprite.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

#pragma mark SFTouchSprite

SFTouchSprite::SFTouchSprite(void)
: SFSprite()
, m_pController(NULL)
, m_bIrregular(false)
, m_tTouchBounds(CCRectZero)
{
}

SFTouchSprite::~SFTouchSprite(void)
{
	setController(NULL);
}

bool SFTouchSprite::init(void)
{
	if (!SFSprite::init())
	{
		CNLog("error");
		return false;
	}
	
	if (!m_pController) // no need to create the touch controller again
	{
		CNTouchController * pController = new CNTouchController();
		if (pController && pController->init())
		{
			setController(pController);
			CC_SAFE_RELEASE(pController);
		}
		else
		{
			CNLog("could not happen!");
			CC_SAFE_RELEASE(pController);
			return false;
		}
	}
	
	setIrregular(false);
	setTouchBounds(CCRectZero);
	
	return true;
}

void SFTouchSprite::appointHandler(CCDictionary * pEventTree, CCDictionary * pActionTree)
{
	SFTouchHandler * pHandler = SFTouchHandler::handler(pEventTree, pActionTree);
	setEventDelegate(pHandler);
}

void SFTouchSprite::onEnter()
{
	SFSprite::onEnter();
	
	if (m_pController)
	{
		m_pController->setNode(this);
		m_pController->setDelegate(dynamic_cast<CNTouchDelegate *>(m_pEventDelegate));
	}
}

void SFTouchSprite::onExit()
{
	if (m_pController)
	{
		m_pController->setDelegate(NULL);
		m_pController->setNode(NULL);
	}
	
	SFSprite::onExit();
}

bool SFTouchSprite::containsNodeSpacePoint(const CCPoint & tPoint)
{
	if (!CCRect::CCRectEqualToRect(m_tTouchBounds, CCRectZero) &&
		!CCRect::CCRectContainsPoint(m_tTouchBounds, tPoint))
	{
		return false;
	}
	return SFSprite::containsNodeSpacePoint(tPoint);
}

bool SFTouchSprite::containsTouchLocation(CCTouch * pTouch)
{
	return CNSprite::containsTouchLocation(pTouch, m_bIrregular);
}

bool SFTouchSprite::setAttributes(CCDictionary * pDict)
{
	if (!SFSprite::setAttributes(pDict))
	{
		CNLog("error");
		return false;
	}
	
	// Irregular
	CCString * pIrregular = (CCString *)pDict->objectForKey("Irregular");
	m_bIrregular = pIrregular ? pIrregular->boolValue() : false;
	
	// TouchBounds
	CCString * pTouchBounds = (CCString *)pDict->objectForKey("TouchBounds");
	m_tTouchBounds = pTouchBounds ? CCRectFromString(pTouchBounds->getCString()) : CCRectZero;
	
	return true;
}

#pragma mark - SFTouchHandler

SFTouchHandler * SFTouchHandler::handler(CCDictionary * pEventTree, CCDictionary * pActionTree)
{
	SFTouchHandler * pHandler = new SFTouchHandler();
	if (pHandler && pHandler->init(pEventTree, pActionTree))
	{
		pHandler->autorelease();
		return pHandler;
	}
	else
	{
		CC_SAFE_RELEASE(pHandler);
		return NULL;
	}
}

void SFTouchHandler::onFingerDown(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onFingerDown", pNode);
	doEvent("onMouseDown", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onFingerUp(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onFingerUp", pNode);
	doEvent("onMouseUp", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onFingerOver(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onFingerOver", pNode);
	doEvent("onMouseOver", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onFingerOut(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onFingerOut", pNode);
	doEvent("onMouseOut", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onFingerMove(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onFingerMove", pNode);
	doEvent("onMouseMove", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

// click
void SFTouchHandler::onClick(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onClick", pNode);
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onDoubleClick(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onDoubleClick", pNode); // deprecated
	CCPoolManager::sharedPoolManager()->pop();
}

// guestures
void SFTouchHandler::onSwipeLeft(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onSwipeLeft", pNode);
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onSwipeRight(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onSwipeRight", pNode);
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onSwipeUp(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onSwipeUp", pNode);
	CCPoolManager::sharedPoolManager()->pop();
}

void SFTouchHandler::onSwipeDown(CCNode * pNode)
{
	CCPoolManager::sharedPoolManager()->push();
	doEvent("onSwipeDown", pNode);
	CCPoolManager::sharedPoolManager()->pop();
}

NS_SF_END

File: objc/extension/SFTouchSprite.m
//
//  SFTouchSprite.m
//  SpriteForest2
//
//  Created by Moky on 12-10-3.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFTouchSprite.h"

@implementation SFTouchSprite

@synthesize controller = controller_;
@synthesize irregular = irregular_;
@synthesize touchBounds = touchBounds_;

- (id) init
{
	if ((self = [super init]))
	{
		self.controller = [[[CNTouchController alloc] init] autorelease];
		irregular_ = NO;
		touchBounds_ = CGRectZero;
	}
	return self;
}

- (void) dealloc
{
	[controller_ release];
	
	[super dealloc];
}

- (void) appointHandlerWithEventTree:(NSDictionary *)eTree actionTree:(NSDictionary *)aTree
{
	SFTouchHandler * handler = [SFTouchHandler handlerWithEventTree:eTree actionTree:aTree];
	self.eventHandler = handler;
}

- (void) onEnter
{
	[super onEnter];
	
	controller_.node = self;
	controller_.delegate = (SFTouchHandler *)eventHandler_;
}

- (void) onExit
{
	controller_.delegate = nil;
	controller_.node = nil;
	
	[super onExit];
}

- (BOOL) containsNodeSpacePoint:(CGPoint)point
{
	if (![super containsNodeSpacePoint:point])
	{
		return NO;
	}
	return CGRectEqualToRect(touchBounds_, CGRectZero) || CGRectContainsPoint(touchBounds_, point);
}

- (BOOL) containsTouchLocation:(UITouch *)touch
{
	return [self containsTouchLocation:touch checkPixel:irregular_];
}

- (void) setAttributes:(NSDictionary *)dict
{
	// irregular
	NSString * irregular = [dict objectForKeyCaseInsensitive:@"Irregular"];
	if (irregular != nil)
	{
		self.irregular = [irregular boolValue];
	}
	
	// touch bounds
	NSString * touchBounds = [dict objectForKeyCaseInsensitive:@"TouchBounds"];
	if (touchBounds != nil)
	{
		self.touchBounds = CGRectFromString(touchBounds);
	}
	
	[super setAttributes:dict];
}

@end


@implementation SFTouchHandler

- (void) onFingerDown:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onFingerDown" node:node];
	[self doEvent:@"onMouseDown" node:node]; // deprecated
	[pool release];
}

- (void) onFingerUp:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onFingerUp" node:node];
	[self doEvent:@"onMouseUp" node:node]; // deprecated
	[pool release];
}

- (void) onFingerOver:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onFingerOver" node:node];
	[self doEvent:@"onMouseOver" node:node]; // deprecated
	[pool release];
}

- (void) onFingerOut:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onFingerOut" node:node];
	[self doEvent:@"onMouseOut" node:node]; // deprecated
	[pool release];
}

- (void) onFingerMove:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onFingerMove" node:node];
	[self doEvent:@"onMouseMove" node:node]; // deprecated
	[pool release];
}

// click
- (void) onClick:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onClick" node:node];
	[pool release];
}

- (void) onDoubleClick:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onDoubleClick" node:node]; // deprecated
	[pool release];
}

// guestures
- (void) onSwipeLeft:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onSwipeLeft" node:node];
	[pool release];
}

- (void) onSwipeRight:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onSwipeRight" node:node];
	[pool release];
}

- (void) onSwipeUp:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onSwipeUp" node:node];
	[pool release];
}

- (void) onSwipeDown:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onSwipeDown" node:node];
	[pool release];
}

@end

